### ISSUE TITLE
404 Page Not Found when trying to access /api/generate

### ISSUE THREAD

**Issue by Sunny-DCosta** [ollama/ollama#10016]
I am running Ollama version 0.6.2 on Windows 10. When I try to access the /api/generate endpoint using curl, I receive a 404 Page Not Found error. I have confirmed that Ollama is running and listening on port 11434, but the specific endpoint seems to be unavailable.

Steps taken:
1. Started Ollama using the command: ollama start
2. Confirmed Ollama is running by accessing the root endpoint: curl http://127.0.0.1:11434/
3. Attempted to access /api/generate: curl http://127.0.0.1:11434/api/generate

Error message:
ERROR - [test2.py:383] - Processing error: HTTPConnectionPool(host='localhost', port=11434): Max retries exceeded with url: /api/generate (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x00000190D9F4E350>: Failed to establish a new connection: [WinError 10061] No connection could be made because the target machine actively refused it'))
Any help would be appreciated.

**Comment by pdevine** [comment]
@Sunny-DCosta please try @rick-github 's `curl` command above. Can you post the code which you're trying to run in `test2.py` which is having the issue? Is it using ollama's python bindings?

**Comment by Sunny-DCosta** [comment]
> [@Sunny-DCosta](https://github.com/Sunny-DCosta) please try [@rick-github](https://github.com/rick-github) 's `curl` command above. Can you post the code which you're trying to run in `test2.py` which is having the issue? Is it using ollama's python bindings?

import logging
import shutil
import os
import streamlit as st
from langchain_community.embeddings import OllamaEmbeddings
from langchain_community.llms import Ollama
from langchain_chroma import Chroma
from langchain.docstore.document import Document
from datasets import load_dataset, concatenate_datasets
from chromadb.config import Settings
from typing import List, Dict, Any, Tuple
import re
from transformers import AutoTokenizer, AutoModelForSeq2SeqLM, AutoModelForVision2Seq, AutoProcessor
import torch
from concurrent.futures import ThreadPoolExecutor, as_completed
import time
import json
import functools
from duckduckgo_search import DDGS
from dataclasses import dataclass
from enum import Enum
import datetime
from PIL import Image
import requests
from io import BytesIO

# Configuration
CHROMA_DIR = os.path.abspath("chroma_db")
METADATA_FILE = os.path.abspath("./vectorstore_metadata.json")
VERSION = "1.0"
SAMPLES_PER_CATEGORY = 2500
MODEL_CACHE_DIR = os.path.abspath("./model_cache")

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(levelname)s - [%(filename)s:%(lineno)d] - %(message)s',
    handlers=[
        logging.FileHandler('rag_app.log'),
        logging.StreamHandler()
    ]
)
logger = logging.getLogger(__name__)

def log_time(func):
    """Decorator to log function execution time"""
    @functools.wraps(func)
    def wrapper(*args, **kwargs):
        start_time = time.time()
        logger.info(f"Starting {func.__name__}")
        result = func(*args, **kwargs)
        end_time = time.time()
        logger.info(f"Completed {func.__name__} in {end_time - start_time:.2f} seconds")
        return result
    return wrapper

def save_metadata():
    """Save metadata about the vectorstore"""
    metadata = {
        "version": VERSION,
        "creation_date": time.strftime("%Y-%m-%d %H:%M:%S"),
        "num_documents": st.session_state.get("num_documents", 0),
        "categories": ["World", "Sports", "Business", "Science/Technology"]
    }
    try:
        with open(METADATA_FILE, "w") as f:
            json.dump(metadata, f)
        logger.info(f"Saved vectorstore metadata to {METADATA_FILE}")
    except Exception as e:
        logger.error(f"Error saving metadata: {str(e)}")

def load_metadata() -> dict:
    """Load vectorstore metadata"""
    try:
        with open(METADATA_FILE, "r") as f:
            metadata = json.load(f)
        logger.info(f"Loaded vectorstore metadata: {metadata}")
        return metadata
    except FileNotFoundError:
        logger.info("No existing metadata found")
        return None

def check_vectorstore_exists() -> Tuple[bool, str]:
    """Check if a valid vectorstore exists"""
    if not (os.path.exists(CHROMA_DIR) and os.path.exists(METADATA_FILE)):
        return False, "Vectorstore not found"
    metadata = load_metadata()
    if not metadata or metadata.get("version") != VERSION:
        return False, "Vectorstore version mismatch"
    return True, "Valid vectorstore found"

def preprocess_question(question: str) -> str:
    """Clean and standardize the input question"""
    return ' '.join(question.lower().strip().split())

@log_time
def extract_entities(text: str, model, tokenizer, num_entities=5) -> List[str]:
    """Extract entities from text with improved accuracy"""
    input_text = f"Extract only the most important entities (people, companies, topics) from: {text}"
    input_ids = tokenizer(input_text, return_tensors="pt", max_length=512, truncation=True).input_ids
    outputs = model.generate(input_ids, max_length=50, num_return_sequences=1, temperature=0.3)
    entities = tokenizer.decode(outputs[0], skip_special_tokens=True).split(',')
    entities = [e.strip() for e in entities if e.strip() and len(e.strip()) > 2]
    question_words = text.lower().split()
    important_words = [word for word in question_words if
                       len(word) > 3 and word not in ['what', 'when', 'where', 'how', 'why', 'who', 'were', 'from']]
    entities.extend(important_words)
    seen = set()
    return [e for e in entities if not (e in seen or seen.add(e))][:num_entities]

def generate_prompt(question: str, context: str, entities: List[str]) -> str:
    """Generate improved prompt with strict formatting rules"""
    if "date" in question.lower() or "time" in question.lower():
        now = datetime.datetime.now()
        return f"""Current UTC Date and Time: 
        Date: {now.strftime('%B %d, %Y')}
        Time: {now.strftime('%I:%M %p')} UTC+0
        """

    if not context:
        return f"""The question "{question}" doesn't appear to have valid information available. 
        Please respond that the requested information couldn't be verified through available sources."""

    if any(kw in question.lower() for kw in ["price", "stock", "value"]):
        return f"""Extract numerical financial data from this context. Format as:
        - Company/Asset: [Name]
        - Current Price: [Number]
        - Currency: [USD/etc]
        - As of: [Date]
        - Sources: [List]
        Context: {context}"""

    return f"""Based on these verified sources: {context}
    Answer: {question}
    Requirements:
    - If information is unavailable, state "No verified information found"
    - For numerical data, include exact numbers and dates
    - Highlight contradictions between sources
    - Format using clear bullet points"""

@log_time
def process_batch(batch, embeddings, chroma_settings, batch_num, total_batches):
    """Process document batches with error handling"""
    try:
        logger.info(f"Processing batch {batch_num}/{total_batches}")
        vectorstore = Chroma.from_documents(
            documents=batch,
            embedding=embeddings,
            persist_directory=CHROMA_DIR,
            client_settings=chroma_settings
        )
        return len(batch)
    except Exception as e:
        logger.error(f"Error in batch {batch_num}/{total_batches}: {str(e)}")
        return 0

@log_time
def initialize_vectorstore(force_reinit: bool = False):
    """Initialize or load vectorstore"""
    if force_reinit:
        shutil.rmtree(CHROMA_DIR, ignore_errors=True)
        os.remove(METADATA_FILE) if os.path.exists(METADATA_FILE) else None

    exists, message = check_vectorstore_exists()
    if exists and not force_reinit:
        try:
            return Chroma(
                persist_directory=CHROMA_DIR,
                embedding_function=OllamaEmbeddings(model="llama2"),
                client_settings=Settings(anonymized_telemetry=False)
            )
        except Exception as e:
            logger.error(f"Error loading vectorstore: {str(e)}")
            raise

    logger.info("Starting new vectorstore initialization")
    with st.spinner("Initializing knowledge base..."):
        dataset = load_dataset("ag_news", split="train")
        sampled_datasets = [
            dataset.filter(lambda x: x['label'] == cat).shuffle(seed=42).select(range(SAMPLES_PER_CATEGORY))
            for cat in [0, 1, 2, 3]]
        balanced_dataset = concatenate_datasets(sampled_datasets).shuffle(seed=42)

        texts = []
        category_names = {0: "World", 1: "Sports", 2: "Business", 3: "Science/Technology"}
        for idx, item in enumerate(balanced_dataset):
            try:
                title = item['text'].split('.')[0]
                texts.append(Document(
                    page_content=f"{title}. {item['text']}",
                    metadata={"category": category_names[item["label"]], "label": item["label"]}
                ))
            except Exception as e:
                logger.error(f"Error processing document {idx}: {str(e)}")

        embeddings = OllamaEmbeddings(model="llama2")
        batches = [texts[i:i + 50] for i in range(0, len(texts), 50)]

        with ThreadPoolExecutor(max_workers=4) as executor:
            list(executor.map(
                lambda b: process_batch(b[1], embeddings, Settings(anonymized_telemetry=False), b[0] + 1, len(batches)),
                enumerate(batches)))

        st.session_state.num_documents = len(texts)
        save_metadata()
        return Chroma(persist_directory=CHROMA_DIR, embedding_function=embeddings)

def get_current_date_time():
    return datetime.datetime.now().strftime('%Y-%m-%d %H:%M:%S')

def fetch_date_time_query(query):
    if 'time' in query.lower() or 'date' in query.lower():
        return get_current_date_time()
    return "Query does not contain date or time request."

def format_answer(text: str) -> str:
    """Format the answer with smart formatting"""
    if "Date:" in text and "Time:" in text:
        return text.split('\n')[0].strip()
    if '\n' in text or '•' in text:
        return '\n'.join([f"- {line.strip()}" for line in text.split('\n') if line.strip()])
    return text

class ExpertType(Enum):
    VECTOR_DB = "vector_db"
    WEB_SEARCH = "web_search"
    HYBRID = "hybrid"

@dataclass
class SearchResult:
    content: str
    source: str
    score: float = 0.0
    metadata: Dict[str, Any] = None

class WebSearchExpert:
    """Improved web search expert with better result handling and multiple loops"""

    def __init__(self, max_results: int = 8, rate_limit: int = 10):
        self.max_results = max_results
        self.rate_limit = rate_limit
        self.last_search_time = time.time()

    def search(self, query: str) -> List[SearchResult]:
        current_time = time.time()
        if current_time - self.last_search_time < 1 / self.rate_limit:
            time.sleep(1 / self.rate_limit - (current_time - self.last_search_time))
        self.last_search_time = current_time

        try:
            with DDGS() as ddgs:
                results = []
                seen_urls = set()

                for i in range(5):
                    modified_query = self._modify_query(query, i)
                    for result in ddgs.text(modified_query, max_results=self.max_results):
                        if result.get('href') not in seen_urls:
                            content = f"{result.get('title', '')}. {result.get('body', '')}"
                            source = result.get('href', 'No Link')
                            results.append(SearchResult(content=content, source=source))
                            seen_urls.add(result.get('href'))
                return self._filter_results(results, query)
        except Exception as e:
            logger.error(f"Web search error: {str(e)}")
            return []

    def _modify_query(self, query: str, iteration: int) -> str:
        modifiers = [
            "latest", "current", "2024 update", "recent information", "accurate"
        ]
        return f"{query} {modifiers[iteration % len(modifiers)]}" if iteration > 0 else query

    def _filter_results(self, results: List[SearchResult], query: str) -> List[SearchResult]:
        query_keywords = set(re.findall(r'\w+', query.lower()))
        filtered = []

        for result in results:
            content = result.content.lower()
            matches = sum(1 for kw in query_keywords if kw in content)
            if any(domain in result.source for domain in ['bloomberg', 'reuters', 'yahoo', 'nasdaq']):
                result.score += 1.0
            if matches > 0:
                result.score += matches * 0.2
                filtered.append(result)

        filtered.sort(key=lambda x: x.score, reverse=True)
        return filtered[:self.max_results * 2]

class VectorDBExpert:
    def __init__(self, vectorstore, k: int = 5):
        self.vectorstore = vectorstore
        self.k = k

    def search(self, query: str) -> List[SearchResult]:
        try:
            return [
                SearchResult(doc.page_content, "vectorstore", score, doc.metadata)
                for doc, score in self.vectorstore.similarity_search_with_score(query, k=self.k)
            ]
        except Exception as e:
            logger.error(f"Vector search error: {str(e)}")
            return []

class ExpertRouter:
    def __init__(self, llm):
        self.llm = llm

    def classify_question(self, question: str) -> ExpertType:
        prompt = f"""Classify this question as VECTOR_DB, WEB_SEARCH, or HYBRID: "{question}" 
        Respond with only one word."""
        try:
            response = self.llm(prompt).strip().upper()
            return ExpertType(response.lower())
        except:
            return ExpertType.HYBRID

class EnhancedRAG:
    def __init__(self, vectorstore, llm):
        self.router = ExpertRouter(llm)
        self.web_expert = WebSearchExpert()
        self.vector_expert = VectorDBExpert(vectorstore)
        self.llm = llm
        self.tokenizer = None
        self.model = None
        self.vlm_model = None
        self.vlm_processor = None

    def set_entity_extraction_model(self, tokenizer, model):
        self.tokenizer = tokenizer
        self.model = model

    def set_vlm_model(self, vlm_model, vlm_processor):
        self.vlm_model = vlm_model
        self.vlm_processor = vlm_processor

    def process_question(self, question: str, uploaded_image: bytes = None) -> Dict[str, Any]:
        try:
            if "date" in question.lower() or "time" in question.lower():
                return {
                    "answer": generate_prompt(question, "", []),
                    "sources": [],
                    "expert_type": "system",
                    "entities": []
                }

            expert_type = self.router.classify_question(question)
            results = []

            if expert_type in [ExpertType.WEB_SEARCH, ExpertType.HYBRID]:
                web_results = self.web_expert.search(question)
                results.extend(web_results)

            if expert_type in [ExpertType.VECTOR_DB, ExpertType.HYBRID]:
                results.extend(self.vector_expert.search(question))

            context = "\n".join([f"[{r.source}] {r.content}" for r in results[:8]])

            if uploaded_image:
                logger.info("Processing uploaded image")
                try:
                    image = Image.open(BytesIO(uploaded_image))
                    inputs = self.vlm_processor(images=image, return_tensors="pt")
                    outputs = self.vlm_model.generate(**inputs)
                    image_description = self.vlm_processor.decode(outputs[0], skip_special_tokens=True)
                    context += f"\nImage description: {image_description}"
                    logger.info(f"Image description generated: {image_description}")
                except Exception as e:
                    logger.error(f"Error processing uploaded image: {str(e)}")
                    return {"answer": "Error processing the uploaded image. Please try again.", "sources": []}

            entities = extract_entities(question, self.model, self.tokenizer) if self.tokenizer else []
            valid_entities = [e for e in entities if any(e.lower() in context.lower() for e in entities)]

            answer = self.llm(generate_prompt(question, context, valid_entities))
            return {
                "answer": self._clean_answer(answer),
                "sources": [{"content": r.content[:200] + "...", "source": r.source} for r in results[:3]],
                "expert_type": expert_type.value,
                "entities": valid_entities
            }
        except Exception as e:
            logger.error(f"Processing error: {str(e)}")
            return {"answer": "Error processing question", "sources": []}

    def _clean_answer(self, text: str) -> str:
        """Remove markdown and invalid formatting"""
        text = re.sub(r'\*+', '', text)
        text = re.sub(r'\n\s*\n', '\n\n', text)
        return text.strip()

def save_feedback(question: str, answer: str, feedback: str, expected_answer: str = None):
    """Save user feedback to a file"""
    feedback_file = "user_feedback.json"
    feedback_data = {
        "question": question,
        "answer": answer,
        "feedback": feedback,
        "expected_answer": expected_answer,
        "timestamp": datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S")
    }
    try:
        with open(feedback_file, "a") as f:
            json.dump(feedback_data, f)
            f.write("\n")
        logger.info(f"Feedback saved to {feedback_file}")
    except Exception as e:
        logger.error(f"Error saving feedback: {str(e)}")

def main():
    logger.info("Starting Enhanced RAG application")
    st.title("Enhanced RAG Application with Web Search")

    if 'initialized' not in st.session_state:
        st.session_state.initialized = False

    with st.sidebar:
        st.header("Admin Controls")
        if st.button("Force Reinitialization"):
            shutil.rmtree(CHROMA_DIR, ignore_errors=True)
            os.remove(METADATA_FILE) if os.path.exists(METADATA_FILE) else None
            st.session_state.initialized = False
            st.success("Vectorstore reset. Refresh to reinitialize.")
            st.stop()

    if not st.session_state.initialized:
        try:
            st.session_state.vectorstore = initialize_vectorstore()
            st.session_state.llm = Ollama(model="llama2")
            model_name = "google/flan-t5-base"
            st.session_state.tokenizer = AutoTokenizer.from_pretrained(model_name)
            st.session_state.model = AutoModelForSeq2SeqLM.from_pretrained(model_name)
            vlm_model_name = "Salesforce/blip-image-captioning-large"
            st.session_state.vlm_model = AutoModelForVision2Seq.from_pretrained(vlm_model_name, cache_dir=MODEL_CACHE_DIR)
            st.session_state.vlm_processor = AutoProcessor.from_pretrained(vlm_model_name, cache_dir=MODEL_CACHE_DIR)
            st.session_state.enhanced_rag = EnhancedRAG(st.session_state.vectorstore, st.session_state.llm)
            st.session_state.enhanced_rag.set_entity_extraction_model(st.session_state.tokenizer,
                                                                      st.session_state.model)
            st.session_state.enhanced_rag.set_vlm_model(st.session_state.vlm_model,
                                                        st.session_state.vlm_processor)
            st.session_state.initialized = True
            st.success("✅ System ready!")
        except Exception as e:
            st.error(f"Initialization failed: {str(e)}")
            return

    question = st.text_input("Enter your question:")
    uploaded_image = st.file_uploader("Upload an image (optional):", type=["jpg", "jpeg", "png"])
    submit_question = st.button("Submit Question")

    if submit_question and question and st.session_state.initialized:
        try:
            with st.spinner("Processing..."):
                result = st.session_state.enhanced_rag.process_question(question, uploaded_image.read() if uploaded_image else None)
                st.write("### Answer")
                st.markdown(result["answer"])

                if result["sources"]:
                    st.write("### Sources")
                    for src in result["sources"]:
                        st.write(f"- [{src['source']}] {src['content']}")

                feedback = st.text_input("Was the answer helpful? (yes/no):")
                expected_answer = st.text_input("If not, please provide the expected answer (optional):")
                submit_feedback = st.button("Submit Feedback")

                if submit_feedback and feedback:
                    save_feedback(question, result["answer"], feedback, expected_answer)
                    logger.info(f"User feedback saved: {feedback}")
        except Exception as e:
            st.error(f"Error: {str(e)}")

if __name__ == "__main__":
    main()

**Comment by Sunny-DCosta** [comment]
Well I do not know how but the problem has resolved, thank you for your aid @rick-github and @pdevine . Could I maybe reach out to you for other issues not related exactly to ollama? I am currently creating a RAG project and am in dire need of guidance. No AI is able to resolve my issues and guide me.
